paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java

```diff
@@ -130,9 +135,23 @@ public TableWriteImpl<InternalRow> newWrite(String commitUser) {
     @Override
     public TableWriteImpl<InternalRow> newWrite(
             String commitUser, ManifestCacheFilter manifestFilter) {
+        AppendOnlyFileStoreWrite writer = store().newWrite(commitUser);
+        boolean negativeBucket = coreOptions().bucket() == -1;
+        if (negativeBucket) {
+            writer.skipCompaction();
+            writer.fromEmptyWriter();
+        }
         return new TableWriteImpl<>(
-                store().newWrite(commitUser, manifestFilter),
-                new InternalRowKeyAndBucketExtractor(tableSchema),
+                writer,
+                new InternalRowKeyAndBucketExtractor(tableSchema) {
+                    @Override
+                    public int bucket() {
+                        if (negativeBucket) {
```

Comment:
We should have a `BucketProcessor` to deal with all bucket related things for source and sink.
You can finish this in next PR.